File: components/autofill_assistant/browser/chip.cc
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/autofill_assistant/browser/chip.h"
#include "components/autofill_assistant/browser/user_action.h"

namespace autofill_assistant {

Chip::Chip() = default;
Chip::~Chip() = default;
Chip::Chip(const Chip& other) = default;
Chip::Chip(const ChipProto& proto)
    : type(proto.type()),
      icon(proto.icon()),
      text(proto.text()),
      sticky(proto.sticky()),
      content_description(proto.content_description()),
      is_content_description_set(proto.has_content_description()) {}

bool Chip::empty() const {
  return type == UNKNOWN_CHIP_TYPE && text.empty() && icon == NO_ICON;
}

void SetDefaultChipType(std::vector<UserAction>* user_actions) {
  for (UserAction& user_action : *user_actions) {
    if (user_action.chip().empty())
      continue;

    if (user_action.chip().type == UNKNOWN_CHIP_TYPE) {
      // Assume chips with unknown type are normal actions.
      user_action.chip().type = NORMAL_ACTION;
    }
  }
}

}  // namespace autofill_assistant

File: chrome/android/java/src/org/chromium/chrome/browser/download/home/filter/chips/Chip.java
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.download.home.filter.chips;

import androidx.annotation.DrawableRes;
import androidx.annotation.StringRes;

import org.chromium.ui.widget.ChipView;

/**
 * A generic visual representation of a Chip. Most of the visuals are immutable, but the selection
 * and enable states are not.
 */
public class Chip {
    /** An id to use for {@link #icon} when there is no icon on the chip. */
    public static final int INVALID_ICON_ID = ChipView.INVALID_ICON_ID;

    /** An id used to identify this chip. */
    public final int id;

    /** The resource id for the text to show in the chip. */
    public final @StringRes int text;

    /** The accessibility text to use for the chip. */
    public String contentDescription;

    /** The resource id for the icon to use in the chip. */
    public final @DrawableRes int icon;

    /** The {@link Runnable} to trigger when this chip is selected by the UI. */
    public final Runnable chipSelectedListener;

    /** Whether or not this Chip is enabled. */
    public boolean enabled;

    /** Whether or not this Chip is selected. */
    public boolean selected;

    /** Builds a new {@link Chip} instance.  These properties cannot be changed. */
    public Chip(int id, @StringRes int text, @DrawableRes int icon, Runnable chipSelectedListener) {
        this.id = id;
        this.text = text;
        this.icon = icon;
        this.chipSelectedListener = chipSelectedListener;
    }
}